### PR TITLE
Credential.tokenStateFlow

### LINKED ISSUE #210: Expose a Credential Token in a Coroutines StateFlow
### Describe the feature request?

Modern UI frameworks are built around observable data (`LiveData`, `StateFlow`, `State` in Jetpack Compose). Currently none of the objects within the SDK expose their data in an observable form. This means that external logic needs to be implemented to cause UIs to manually refresh their snapshot of data when state within the Okta SDK is updated.

This feature request is basically requesting that the `Credential` object exposes its current token via a `StateFlow` which would allow UIs to monitor the token for changes and automatically refresh their UI.

The reason I'm asking for access to the `token` in an observable format is to automatically drive a UI change when the user is logged in vs logged out.

While you are at it, it might be beneficial to also expose the `tags` via a `StateFlow` as well.

### New or Affected Resource(s)

This should only affect the `Credential` object.

### Provide a documentation link

_No response_

### Additional Information?

I'd imagine this could be implemented via storing the private `_token` var as a `MutableStateFlow<Token?>`.
All access to it can then be accomplished internally using `_token.value`.
And a new property could be exposed: `val tokenStateFlow: StateFlow<Token?> get() = _token`

### PR DESCRIPTION
**Pull request by frett**
This exposes the current token for a `Credential` object as a `StateFlow` that can be observed.

Resolves #210

### CODE REVIEW

**Review by JayNewstrom — COMMENTED** [review]
You'll need a few things before this can be merged:
- Run `./gradlew apiDump` to get the latest public APIs updated 
- Add tests for the new `val`.
- The flow should be closed when the credential `delete` method is called.

**Comment by JayNewstrom** [comment]
I guess StateFlow isn't closable. I'm not sure how to signal this when the credential is deleted. How do we signal through to the flow they need to look up a new credential?

**Comment by frett** [comment]
Hmm, a StateFlow can't be closed according to the API docs. Their suggestion in the API docs is:
> All errors and completion signals should be explicitly materialized if needed.

Possible solutions I see:
- create a sealed wrapper object to reflect the state
- set the value to null when it's deleted, with no further updates to it (this would mirror the `token` property behavior of just returning null when the credential is deleted)
- expose a `Flow` instead of a `StateFlow`, which could be closed.

**Comment by frett** [comment]
I think my preference would be expose it as a `Flow` that can be closed. Any preference from the API standpoint since it is your API to maintain long term ;)

**Comment by JayNewstrom** [comment]
Yeah, I think I'd prefer to expose it as a `Flow`. A consumer can convert it to a `StateFlow`, using the `token` property, if they require it. But I'd like to expose it in a closable way.

**Comment by frett** [comment]
I'll work up those changes now, I'm going to probably implement it as an internal sealed state hierarchy so that internal state stays consistent and atomic.

**Comment by JayNewstrom** [comment]
One other thing, we'll also need you to sign the CLA https://developer.okta.com/cla/

File: auth-foundation/src/main/java/com/okta/authfoundation/credential/Credential.kt
Comment (frett) [inline auth-foundation/src/main/java/com/okta/authfoundation/credential/Credential.kt:90]: this will emit a final null value before closing the `Flow`, I can rework this a bit if we don't want that final null emitted.

Comment (frett) [reply auth-foundation/src/main/java/com/okta/authfoundation/credential/Credential.kt:90]: at least that is my understanding of how `transformWhile` works, this is my first time using it.

**Comment by frett** [comment]
> One other thing, we'll also need you to sign the CLA https://developer.okta.com/cla/

just emailed the CLA. going to work on writing some tests now

**Comment by frett** [comment]
@JayNewstrom thoughts on using the turbine library to test out the flow behavior?

https://github.com/cashapp/turbine

**Comment by JayNewstrom** [comment]
Go for it 👍

File: auth-foundation/src/test/java/com/okta/authfoundation/credential/CredentialTest.kt
Comment (JayNewstrom) [inline auth-foundation/src/test/java/com/okta/authfoundation/credential/CredentialTest.kt:153]: This case seems like it should instantly be closed, and not return any items.

**Review by JayNewstrom — COMMENTED** [review]
Overall this looks good. Thanks for putting in the work for this.

File: auth-foundation/src/test/java/com/okta/authfoundation/credential/CredentialTest.kt
Comment (frett) [reply auth-foundation/src/test/java/com/okta/authfoundation/credential/CredentialTest.kt:153]: k, I do think it should emit a final `null` value for anyone already observing a token flow.

I'll see what I can come up with to return an empty flow if the credential was already deleted without changing the final `null` behavior

**Comment by JayNewstrom** [comment]
@frett I typically do a `./gradlew apiDump spotlessApply testDebugUnitTest` to make sure everything is good locally. (Looks like lint is failing right now).

File: auth-foundation/src/main/java/com/okta/authfoundation/credential/Credential.kt
Comment (frett) [inline auth-foundation/src/main/java/com/okta/authfoundation/credential/Credential.kt:105]: @JayNewstrom I ended up going with this instead of a simple `if (isDeleted) return emptyFlow()` because this avoids a minor race condition if the credential is deleted after that block before the actual flow is returned.

It was definitely an edge case that would give a flow with a single null before closing.

Comment (frett) [reply auth-foundation/src/main/java/com/okta/authfoundation/credential/Credential.kt:105]: I don't feel that strongly about accidentally emitting a single element flow of `null`, so I can switch it back if you would prefer the simplicity of that solution over a solution without the race condition

**Comment by frett** [comment]
> @frett I typically do a `./gradlew apiDump spotlessApply testDebugUnitTest` to make sure everything is good locally. (Looks like lint is failing right now).

was just an unused import, just force pushed the fixup

**Comment by JayNewstrom** [comment]
Everything looks good. Thanks again for the PR!

I'm going to take the weekend to mull over this to make sure I'm not missing anything, and I also need to wait on confirmation of the CLA from an internal team member.

**Comment by frett** [comment]
Sounds good, do you have a rough timeline for when you will be publishing a new version? or is there a SNAPSHOT build published anywhere I can consume until a new release comes out? (obviously the snapshot won't include this until it gets merged)

**Comment by JayNewstrom** [comment]
I'm hoping to get a released published next week. But that's not a promise! One of my coworkers is working on another feature we're hoping to land in 1.1!

We don't publish snapshots right now. You can always use something like gradle composite builds to pull it in locally to do development against, but that's likely not going to be worth setting up in your CI to be able to merge with it.

**Review by JayNewstrom — APPROVED** [review]
Thanks again for the PR!

**Pull request merged**
